src/kerberos.c

```diff
@@ -211,14 +211,56 @@ static PyObject *authGSSClientClean(PyObject *self, PyObject *args)
     return Py_BuildValue("i", result);
 }
 
-static PyObject *authGSSClientStep(PyObject *self, PyObject *args)
+static PyObject *buildChannelBindingsStruct(PyObject *self, PyObject *args, PyObject* keywds)
+{
+    int initiator_addrtype = GSS_C_AF_UNSPEC;
+    int acceptor_addrtype = GSS_C_AF_UNSPEC;
+    int initiator_length = 0;
+    int acceptor_length = 0;
+    int application_length = 0;
+    const char *initiator_address = NULL;
+    const char *acceptor_address = NULL;
+    const char *application_data = NULL;
+    int result = 0;
+
+    PyObject *pychan_bindings = NULL;
+    struct gss_channel_bindings_struct *input_chan_bindings;
+	static char *kwlist[] = {"initiator_addrtype", "initiator_address", "accept_addrtype",
+	                        "acceptor_address", "application_data", NULL};
+
+	if (!PyArg_ParseTupleAndKeywords(args, keywds, "|is#is#s#", kwlist,
+	        &initiator_addrtype, &initiator_address, &initiator_length,
+	        &acceptor_addrtype, &acceptor_address, &acceptor_length,
+	        &application_data, &application_length)) {
+	    return NULL;
+	}
+
+	input_chan_bindings = (struct gss_channel_bindings_struct *) malloc(sizeof(struct gss_channel_bindings_struct));
+	pychan_bindings = PyCObject_FromVoidPtr(input_chan_bindings, free);
```

Comment:
In Python 3.x, PyCObject_FromVoidPtr is an alias for PyCapsule_New (see line 40). The destructor passed as the second argument here will end up being called on the PyCapsule itself, not the gss_channel_bindings_struct. That's almost certainly not what you want.

https://docs.python.org/3.1/c-api/capsule.html#PyCapsule_New